### ISSUE TITLE
FTP timeout

### ISSUE THREAD

**Issue by daaru00** [Equals182/eqFTP#96]
if I do not load files for a few minutes i can't upload any file unless relaunch brackets

**Comment by f34rl3ss** [comment]
Same problem here, not sure how long the interval is before this happens (i would assume like 30min).
Let's say i take a break and come back 30min later, brackets stayed open the whole time, i change a file and want to upload it, i get a red error message from eqFTP, reading "FTP Error: EPIPE".
After that, i cannot upload or reconnect to the server. 
All i can do is:
1) close brackets completly and reopen
2) push "F5" so brackets does a refresh
After that, i can connect to the server again.

Log after first encounter of error:
``` 
[node-error 4:33:19 PM] [eqFTP-ftpDomain]: [s.l][FTP-error] {"code":"EPIPE"}
main.js:1644 Object {event: "server_connection_error", err: Object, connectionID: 0}
main.js:1644 Object {event: "server_disconnect", err: Object, connectionID: 0}
main.js:1859 [eqFTP] Disconnected from server
NodeDebugUtils.js:114 [node-error 4:33:19 PM] [eqFTP-ftpDomain]: [q.p] Can't connect to server on connectionID: 0. Queue cleared now.
NodeDebugUtils.js:117 [node-log 4:33:19 PM] [eqFTP-ftpDomain]: Disconnecting...
main.js:1644 Object {event: "server_disconnecting", connectionID: 0, clearQueue: true}
NodeDebugUtils.js:117 [node-log 4:33:19 PM] [eqFTP-queueisbusy][c.d no connection] Setting busy to false
NodeDebugUtils.js:117 [node-log 4:33:19 PM] [eqFTP-ftpDomain]: Disonnected. Actually there was no client so nothing to disconnect.
NodeDebugUtils.js:117 [node-log 4:33:19 PM] [eqFTP-rawTest] doQueue. command is probably sent. Fuck there's a lot to debug...
NodeDebugUtils.js:117 [node-log 4:33:19 PM] DEBUG:  user_command
NodeDebugUtils.js:117 [node-log 4:33:19 PM] "pwd"
```
when trying to save again the log shows:
```
 GET file:///F:/Files/jic.li/wp-content/themes/jic/images/navi_texture_2.jpg net::ERR_FILE_NOT_FOUND
main.js:3295 16:39:0.462 [eqFTP-uploadonsave] documentSaved triggered
main.js:3297 [eqFTP-uploadonsave] Checking if file is within project F:/Files/jic.li/wp-content/themes/jic/style_jic.css
main.js:3299 [eqFTP-uploadonsave] Yes it is
main.js:3301 [eqFTP-uploadonsave] Getting connection ID by path
main.js:3303 [eqFTP-uploadonsave] Here it is:  0
main.js:3306 [eqFTP-uploadonsave] Automatization type is classic
main.js:3310 [eqFTP-uploadonsave] Upload on save is on, everything is ok
main.js:3323 [eqFTP-uploadonsave] Here's queuer:  Object {localPath: "F:/Files/jic.li/wp-content/themes/jic/style_jic.css", remotePath: "/wp-content/themes/jic/style_jic.css", name: "style_jic.css", direction: "upload", queue: "a"…}
main.js:3326 [eqFTP-uploadonsave] Adding to queue.
main.js:1595 16:39:0.464 Adding to queue Object {localPath: "F:/Files/jic.li/wp-content/themes/jic/style_jic.css", remotePath: "/wp-content/themes/jic/style_jic.css", name: "style_jic.css", direction: "upload", queue: "a"…}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [s.d.i.cs] Checking for ignoring.
main.js:1644 Object {event: "queue_update", connectionID: 0, element: Object}
main.js:1644 Object {event: "working", work: true}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: 1887: [q.a] Queuer is NOT ignored, continueing
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: 1918: [q.a] Queue updated: {"a":[{"localPath":"F:/Files/jic.li/wp-content/themes/jic/style_jic.css","remotePath":"/wp-content/themes/jic/style_jic.css","name":"style_jic.css","direction":"upload","queue":"a","type":"file","connectionID":0,"after":"disconnect","id":"f_143507034046483"}],"p":[],"f":[],"s":[]}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-queueisbusy][q.p just started] Setting busy to true
main.js:1644 Object {event: "queue_update", connectionID: 0, element: Object}
main.js:1644 Object {event: "server_connecting", connectionID: 0}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: Connecting to this ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: Connecting...
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [s.a] Client is already authorized for connection ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [s.gRR] Getting remote root for this ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: Trying to upload file. Local path: F:/Files/jic.li/wp-content/themes/jic/style_jic.css Remote path: /wp-content/themes/jic/style_jic.css
2NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [c.c] Connection already exists for this ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] start
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] step 2
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Command string:  pwd
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Pushing to queue:  { command: 'pwd', callback: [Function], callback100: undefined }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Queue is:  [ { command: { cmd: 'placeholder' }, callback: [Function] },
  { command: 'pwd', callback: [Function], callback100: undefined } ]
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] doing queue
main.js:1644 Object {event: "server_connect", connectionID: 0}
main.js:1920 [eqFTP] Connected to server
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: Connected...
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [s.a] Client is already authorized for connection ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [s.gRR] Getting remote root for this ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: Trying to upload file. Local path: F:/Files/jic.li/wp-content/themes/jic/style_jic.css Remote path: /wp-content/themes/jic/style_jic.css
2NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-ftpDomain]: [c.c] Connection already exists for this ID: 0
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] start
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] step 2
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Command string:  pwd
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Pushing to queue:  { command: 'pwd', callback: [Function], callback100: undefined }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] Queue is:  [ { command: { cmd: 'placeholder' }, callback: [Function] },
  { command: 'pwd', callback: [Function], callback100: undefined },
  { command: 'pwd', callback: [Function], callback100: undefined } ]
NodeDebugUtils.js:117 [node-log 4:39:00 PM] DEBUG:  server_response
NodeDebugUtils.js:117 [node-log 4:39:00 PM] {
  "code": 220,
  "text": "220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------\n220-You are user number 14 of 400 allowed.\n220-Local time is now 16:38. Server port: 21.\n220-This is a private system - No anonymous login\n220-IPv6 connections are also welcome on this server.\n220 You will be disconnected after 15 minutes of inactivity.",
  "isMark": false,
  "isError": false
}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] parsing response { code: 220,
  text: '220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------\n220-You are user number 14 of 400 allowed.\n220-Local time is now 16:38. Server port: 21.\n220-This is a private system - No anonymous login\n220-IPv6 connections are also welcome on this server.\n220 You will be disconnected after 15 minutes of inactivity.',
  isMark: false,
  isError: false }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] no callback { command: { cmd: 'placeholder' }, callback: [Function] }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] calling function (err, data) {
        }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] doQueue. Command is:  pwd
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] doQueue. command is probably sent. Fuck there's a lot to debug...
NodeDebugUtils.js:117 [node-log 4:39:00 PM] DEBUG:  user_command
NodeDebugUtils.js:117 [node-log 4:39:00 PM] "pwd"
NodeDebugUtils.js:117 [node-log 4:39:00 PM] DEBUG:  server_response
NodeDebugUtils.js:117 [node-log 4:39:00 PM] {
  "code": 530,
  "text": "530 You aren't logged in",
  "isMark": false,
  "isError": true
}
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] parsing response { code: 530,
  text: '530 You aren\'t logged in',
  isMark: false,
  isError: true }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] no callback { command: 'pwd', callback: [Function], callback100: undefined }
NodeDebugUtils.js:117 [node-log 4:39:00 PM] [eqFTP-rawTest] calling function (err, data) {
                                            var path = data.text.match(/257\s"(.*?)"/i);
                                            if (!path[1] || path[1] === undefined) {
                                                path = "/";
                                            } else {
                                                path = path[1];
                                            }
                                            if (params.callback)
                                                params.callback(path);
                                            else
                                                return path;
                                        }
NodeDebugUtils.js:114 [node-error 4:39:00 PM] [Launcher] uncaught exception at top level, exiting. TypeError: Cannot read property '1' of null
    at Object.eqFTPconnections.(anonymous function).ftpDomain.client.raw.callback (C:\Users\xxx\AppData\Roaming\Brackets\extensions\user\equals182.eqftp\node\ftpDomain.js:1533:54)
    at Ftp.parseResponse (C:\Users\xxx\AppData\Roaming\Brackets\extensions\user\equals182.eqftp\node\node_modules\jsftp\lib\jsftp.js:256:38)
    at Stream.<anonymous> (C:\Users\xxx\AppData\Roaming\Brackets\extensions\user\equals182.eqftp\node\node_modules\jsftp\lib\jsftp.js:101:14)
    at Stream.EventEmitter.emit (events.js:95:17)
    at ResponseParser.reemit (C:\Users\xxx\AppData\Roaming\Brackets\extensions\user\equals182.eqftp\node\node_modules\jsftp\node_modules\event-stream\node_modules\duplexer\index.js:70:25)
    at ResponseParser.EventEmitter.emit (events.js:95:17)
    at ResponseParser.<anonymous> (_stream_readable.js:746:14)
    at ResponseParser.EventEmitter.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:404:5)
    at readableAddChunk (_stream_readable.js:165:9)
NodeDebugUtils.js:114 [node-info 4:39:00 PM] [Server] stopping
/filesystem/FileSystem.js:957 File watchers went offline!
/utils/NodeConnection.js:89 WebSocket connection to 'ws://localhost:54687/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
/filesystem/FileSystem.js:965 Watching disabled for F:/Files/jic.li/
7/utils/NodeConnection.js:89 WebSocket connection to 'ws://localhost:54687/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
```
when trying to press the disconnect/reconnect button within the eqFTP window:
```
Uncaught TypeError: Cannot read property 'disconnect' of undefined
main.js:1538 eqFTP.ftp.disconnect
main.js:2803 (anonymous function)
thirdparty.min.js:560 n.event.dispatch
thirdparty.min.js:560 r.handle
```
when trying to refresh the file tree within the eqFTP window:
```
16:42:33.737 Adding to queue Object {type: "folder", connectionID: 0, path: "/", id: "f_1435070553737454", state: "opened"…}
main.js:1596 Uncaught TypeError: Cannot read property 'addToQueue' of undefined
main.js:1596 (anonymous function)
thirdparty.min.js:559 n.extend.each
main.js:1589 eqFTP.ftp.queue.add
main.js:1580 eqFTP.ftp.changeDirectory
main.js:2826 (anonymous function)
thirdparty.min.js:560 n.event.dispatch
thirdparty.min.js:560 r.handle
```
Hope this helps :)

**Comment by Equals182** [comment]
The thing is, that server automatically disconnects you after some time. It should send some message before doing that. And *probably* eqFTP handles it wrong. I've tried to solve this issue before and made some progress, but obviously not good enough.

**Comment by 517Design** [comment]
Have you made any progress on this issue or is it fixed? It seems to be a common issue with users. Is there anything I can do? I love brackets and eqftp but it times out every few minutes and Its really frustrating when i have to reload brackets every few minutes or so. It breaks my work flow and is a real time consumer. Thanks, Greg

**Comment by Daaru00** [comment]
My workaround is set "keep alive" property to 0, every time file is uploaded the connection will be reset.

**Comment by Equals182** [comment]
### Hello there!

I know this issue been there for a pretty long time, but I have some good news here!
We just released eqFTP v**1.0.0-beta.0** (v1-b for short I guess) and we're completely discontinuing support of all previous versions of this plugin.
New version works great on Brackets 1.8 and it should be good on lower versions too.

So yeah, we're closing and locking this issue forever and looking forward for you to try out our new version of eqFTP.
I'm sorry if I somehow disturbed you, or you lost your interest in eqFTP or Brackets in general, we're not going to make any noise here again.

If you have any questions, please contact me via email (you can find it in [my profile](https://github.com/Equals182)) or [Telegram app](http://t.me/equals182)

Thank you for understanding 😊
